src/async-wrap.cc

```diff
@@ -177,6 +179,136 @@ static void PushBackDestroyId(Environment* env, double id) {
+  }
+
+  if (async_hooks->fields()[AsyncHooks::kBefore] > 0) {
+    Local<Value> uid = Number::New(wrap->env()->isolate(), wrap->get_id());
+    Local<Function> fn = wrap->env()->async_hooks_before_function();
+    TryCatch try_catch(wrap->env()->isolate());
+    MaybeLocal<Value> ar = fn->Call(
+        wrap->env()->context(), Undefined(wrap->env()->isolate()), 1, &uid);
+    if (ar.IsEmpty()) {
+      ClearFatalExceptionHandlers(wrap->env());
+      FatalException(wrap->env()->isolate(), try_catch);
+      return false;
+    }
+  }
+  return true;
+}
+
+
+static bool PostCallbackExecution(AsyncWrap* wrap) {
+  AsyncHooks* async_hooks = wrap->env()->async_hooks();
+
+  // If the callback failed then the after() hooks will be called at the end
+  // of _fatalException().
+  if (async_hooks->fields()[AsyncHooks::kAfter] > 0) {
+    Local<Value> uid = Number::New(wrap->env()->isolate(), wrap->get_id());
+    Local<Function> fn = wrap->env()->async_hooks_after_function();
+    TryCatch try_catch(wrap->env()->isolate());
+    MaybeLocal<Value> ar = fn->Call(
+        wrap->env()->context(), Undefined(wrap->env()->isolate()), 1, &uid);
+    if (ar.IsEmpty()) {
+      ClearFatalExceptionHandlers(wrap->env());
+      FatalException(wrap->env()->isolate(), try_catch);
+      return false;
+    }
+  }
+
+  if (wrap->env()->using_domains()) {
+    Local<Value> domain_v = wrap->object()->Get(wrap->env()->domain_string());
+    if (domain_v->IsObject()) {
+      Local<Object> domain = domain_v.As<Object>();
+      if (domain->Get(wrap->env()->disposed_string())->IsTrue())
+        return false;
+      Local<Value> exit_v = domain->Get(wrap->env()->exit_string());
+      if (exit_v->IsFunction()) {
+        if (exit_v.As<Function>()->Call(domain, 0, nullptr).IsEmpty()) {
+          FatalError("node::AsyncWrap::MakeCallback",
+                    "domain exit callback threw, please report this");
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+class PromiseWrap : public AsyncWrap {
+ public:
+  PromiseWrap(Environment* env, Local<Object> object)
+    : AsyncWrap(env, object, PROVIDER_PROMISE) {}
+  size_t self_size() const override { return sizeof(*this); }
+};
+
+
+static void GetPromiseDomain(Local<v8::Name> name,
+                             const v8::PropertyCallbackInfo<v8::Value>& info) {
+  Local<Context> context = info.GetIsolate()->GetCurrentContext();
+  Environment* env = Environment::GetCurrent(context);
+  info.GetReturnValue().Set(
+      info.Data().As<Object>()->Get(context,
+                                    env->domain_string()).ToLocalChecked());
+}
+
+
+static void PromiseHook(PromiseHookType type, Local<Promise> promise,
+                        Local<Value> parent, void* arg) {
+  Local<Context> context = promise->CreationContext();
+  Environment* env = Environment::GetCurrent(context);
+  const char async_id_key[] = "__async_wrap";
+  const char tag_id_key[] = "__async_wrap_tag";
+  if (type == PromiseHookType::kInit) {
```

Comment:
@matthewloring I think this might be enough for what you want? I didn’t try it with this PR but it passes testing on its own.

<details><summary>diff</summary>

```diff
diff --git a/lib/async_hooks.js b/lib/async_hooks.js
index 867b5eb52da1..49f8fa5becf2 100644
--- a/lib/async_hooks.js
+++ b/lib/async_hooks.js
@@ -49,12 +49,7 @@ const before_symbol = Symbol('before');
 const after_symbol = Symbol('after');
 const destroy_symbol = Symbol('destroy');
 
-// Setup the callbacks that node::AsyncWrap will call when there are hooks to
-// process. They use the same functions as the JS embedder API.
-async_wrap.setupHooks({ init,
-                        before: emitBeforeN,
-                        after: emitAfterN,
-                        destroy: emitDestroyN });
+let setupHooksCalled = false;
 
 // Used to fatally abort the process if a callback throws.
 function fatalError(e) {
@@ -103,6 +98,16 @@ class AsyncHook {
     if (hooks_array.includes(this))
       return;
 
+    if (!setupHooksCalled) {
+      setupHooksCalled = true;
+      // Setup the callbacks that node::AsyncWrap will call when there are
+      // hooks to process. They use the same functions as the JS embedder API.
+      async_wrap.setupHooks({ init,
+                              before: emitBeforeN,
+                              after: emitAfterN,
+                              destroy: emitDestroyN });
+    }
+
     // createHook() has already enforced that the callbacks are all functions,
     // so here simply increment the count of whether each callbacks exists or
     // not.
```
</details>